This are not working debezium.sink.iceberg.upsert and debezium.sink.iceberg.upsert-keep-deletes and giving us the error.

Hi @Team,

I have been trying to configure Debezium with Iceberg to achieve upserts while ensuring that deleted records are physically removed from the table (not just marked as soft deletes). However, the current configuration is not working as expected.

**First Time :-** 
debezium.sink.type=iceberg
debezium.sink.iceberg.table-prefix=default_table_1
#DEFAULT
#debezium.sink.iceberg.skip-name-validation=true
debezium.sink.iceberg.upsert=true
#false
debezium.sink.iceberg.upsert-keep-deletes=false
debezium.sink.iceberg.write.format.default=parquet
debezium.sink.iceberg.catalog-name=default_catalog_1
debezium.sink.iceberg.table-namespace=tables_debezium_1
debezium.sink.iceberg.destination-lowercase-table-names=true

**Second Time :-** 
debezium.sink.type=iceberg
debezium.sink.iceberg.table-prefix=default_table_1
#DEFAULT
#debezium.sink.iceberg.skip-name-validation=true
debezium.sink.iceberg.upsert=true
#false
debezium.sink.iceberg.upsert-keep-deletes=false
(above one with true is also not working)
#false
debezium.sink.iceberg.write.format.default=parquet
debezium.sink.iceberg.catalog-name=default_catalog_1
debezium.sink.iceberg.table-namespace=tables_debezium_1
debezium.sink.iceberg.destination-lowercase-table-names=true
############################################################################################
debezium.sink.iceberg.write.update.mode=merge-on-read
debezium.sink.iceberg.write.delete.isolation-level=snapshot
debezium.sink.iceberg.write.update.isolation-level=snapshot
debezium.sink.iceberg.write.update.mode=merge-on-read
debezium.sink.iceberg.write.merge.mode=merge-on-read
debezium.sink.iceberg.write.merge.isolation-level=snapshot 


**Issue:**
We want Debezium to handle upserts properly and delete records from Iceberg when they are deleted from the source DB (without soft deletes). However, with the above parameters, this behavior is not working.

Could you please advise on what configurations we need to achieve this?
Also let me share the logs:- 

[Extra_Log_share.log](https://github.com/user-attachments/files/18616824/Extra_Log_share.log)



Thank you,
Dhruv Mevada

@mevadadhruv  error message highlighting the issue. 

`Jan 31 10:22:44 prodldebez01 run.sh[827418]: Caused by: io.debezium.DebeziumException: Field '__source_ts_ms'
not found in JSON object: `

and make sure you have event flattening enabled

https://github.com/memiiso/debezium-server-iceberg/blob/5fe3ea5f35c2ff8447e421eeac8960b47a8a3f56/examples/conf/application.properties#L51-L55

> [@mevadadhruv](https://github.com/mevadadhruv) error message highlighting the issue.
> 
> `Jan 31 10:22:44 prodldebez01 run.sh[827418]: Caused by: io.debezium.DebeziumException: Field '__source_ts_ms' not found in JSON object: `
> 
> and make sure you have event flattening enabled
> 
> [debezium-server-iceberg/examples/conf/application.properties](https://github.com/memiiso/debezium-server-iceberg/blob/5fe3ea5f35c2ff8447e421eeac8960b47a8a3f56/examples/conf/application.properties#L51-L55)
> 
> Lines 51 to 55 in [5fe3ea5](/memiiso/debezium-server-iceberg/commit/5fe3ea5f35c2ff8447e421eeac8960b47a8a3f56)
> 
>  debezium.transforms=unwrap 
>  debezium.transforms.unwrap.type=io.debezium.transforms.ExtractNewRecordState 
>  debezium.transforms.unwrap.add.fields=op,table,source.ts_ms,db 
>  debezium.transforms.unwrap.delete.handling.mode=rewrite 
>  debezium.transforms.unwrap.drop.tombstones=true

Also We are providing this too .
#,changeCase
debezium.transforms.unwrap.type=io.debezium.transforms.ExtractNewRecordState
debezium.transforms.unwrap.add.fields=op,table,source.ts_ms,db
debezium.transforms.unwrap.delete.handling.mode=rewrite
debezium.transforms.unwrap.drop.tombstones=true
#debezium.transforms.changeCase.type=com.github.jcustenborder.kafka.connect.transform.common.ChangeCase$Value
#debezium.transforms.changeCase.from=UPPER_UNDERSCORE
#debezium.transforms.changeCase.to=LOWER_UNDERSCORE

--- a/debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/RecordConverter.java
+++ b/debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/RecordConverter.java
@@ -143,8 +143,8 @@ public class RecordConverter {
     // Schema change events are identified by the presence of "ddl", "databaseName", and "tableChanges" fields.
     // "schema change topic" https://debezium.io/documentation/reference/3.0/connectors/mysql.html#mysql-schema-change-topic
     if (isSchemaChangeEvent()) {
-      LOGGER.warn("Schema change event detected. Creating Iceberg schema without identifier fields for append-only mode.");
-      schemaConverter().icebergSchema(false); // Force no identifier fields for schema changes
+      LOGGER.warn("Schema change topic detected. Creating Iceberg schema without identifier fields for append-only mode.");
+      return schemaConverter().icebergSchema(false); // Force no identifier fields for schema changes
     }
 
     return schemaConverter().icebergSchema(createIdentifierFields);

